### PR TITLE
Refactors AgoraClient's initialization

### LINKED ISSUE #13: [PROPOSAL] Move initialization into it's own function for AgoraClient
In the `AgoraClient`'s constructor there is an implicit call to a initialization function.

https://github.com/AgoraIO-Community/Flutter-UIKit/blob/4ab388978f322caf99234420f5e8f510ebcd0a34/lib/src/agora_client.dart#L32-L44

If you dive more deep you see there are some async stuff going on which makes the constructor unexpectedly heavy.
Other packages have the same need (e.g.: [camera](https://pub.dev/packages/camera) and [video_player](https://pub.dev/packages/video_player)) and do so by giving the client the work of calling `initialize`. It does seem more work from the client's perspective but it's just an one liner and gives the client the power to choose when to kick off the setup.

### PR DESCRIPTION
**Pull request by Ascenio**
Closes #13.

Moves initialization of `AgoraClient` from its constructor (implicit) to a proper function (explicit).
This means that it's the user responsability to check for it, but it also gives the power to know when it's ready.

## Release Notes

- **BREAKING CHANGE:** Must call `initialize` on `AgoraClient` before further usage.

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [ ] The GitHub Actions pass building and linting. Linter returns no warnings or errors.
- [X] The QA checklist below has been completed

## Pull request type

Please check the type of change your PR introduces:
- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?
The initialization is triggered by the constructor of `AgoraClient` at initialization and if it happens to failure, it can't recover from it.

Issue Number: #13

## What is the new behavior?
The initialization is triggered by a specific function by the client. If failure happens the client gets to know and can handle it. Also the same instance can be reused.

## Does this introduce a breaking change?

- [X] Yes
- [ ] No

### UIKit Update Checklist (Minor or Patch Release)

- [X] Using the latest version of Agora's Video SDK
- [X] Example apps are all functional
- [X] Core features are still working (both ways across platforms)
	- [ ] Camera + Mic muting works for local and remote users
	- [ ] Users are added and removed correctly when they join and leave the channel
	- [ ] Older versions of the library gracefully handle changes (Create issue if not)
	- [ ] Builtin buttons all work as expected
- [ ] Any newly deprecated methods are flagged as such inline and in documentation

### CODE REVIEW

File: example/lib/main.dart [hunk 1]
```diff
@@ -11,16 +11,26 @@ class MyApp extends StatefulWidget {
...
+
+  @override
+  void initState() {
+    super.initState();
+    client = AgoraClient(
+      agoraConnectionData: AgoraConnectionData(
+        appId: "9d98d609dca244b6a885e6c9b12a30ba",
+        channelName: "test",
+        tempToken:
+            '0069d98d609dca244b6a885e6c9b12a30baIADRJkUegWzfxt5AtMvEdHqGgqiT+vWfLpA9KgmDpcEZrAx+f9gAAAAAEACqPfBqFZvfYAEAAQAVm99g',
```
Comment (maxxfrazer) [inline example/lib/main.dart:25]: Please revert to original

File: example/lib/main.dart [hunk 2]
```diff
@@ -32,15 +42,27 @@ class _MyAppState extends State<MyApp> {
...
-            ],
+          child: FutureBuilder(
+            future: futureInitialize,
+            builder: (context, snapshot) {
+              if (snapshot.connectionState == ConnectionState.waiting) {
+                return Center(child: CircularProgressIndicator());
+              }
+              if (snapshot.hasError) {
+                return Center(child: Text('Some error occurred.'));
+              }
```
Comment (maxxfrazer) [inline example/lib/main.dart:53]: Could this be more descriptive?

**Review by maxxfrazer — CHANGES_REQUESTED** [review]
Just a couple of edits from me, @Meherdeep and/or @tadaspetra will take a deeper look!

File: example/lib/main.dart (hunk 1 above)
Comment (Ascenio) [reply example/lib/main.dart:25]: Oh my bad :sweat_smile:

File: example/lib/main.dart (hunk 2 above)
Comment (Ascenio) [reply example/lib/main.dart:53]: Sure.

Comment (Ascenio) [reply example/lib/main.dart:53]: I'm thinking on "Coulnd't initialize AgoraClient", that is not something a user from an app would expect but as a demo of the lib I see no problem.

**Comment by tadaspetra** [comment]
Me and @Meherdeep discussed this yesterday, and we think this is the right approach. But we want to go over a couple more things with this including
- Making sure this breaking change doesn't have adverse effects
- Adding dispose as well

**Comment by Ascenio** [comment]
@tadaspetra that is true, I've noticed there was no dispose and that bugs me. Should I include it in this PR? I think the SessionController should be disposed, but I'm not sure in regards of RtcEngine.

**Comment by tadaspetra** [comment]
> @tadaspetra that is true, I've noticed there was no dispose and that bugs me. Should I include it in this PR? I think the SessionController should be disposed, but I'm not sure in regards of RtcEngine.

Of course, we would love it if you could add it 😄

**Comment by Ascenio** [comment]
@tadaspetra I've found out that RtcEngine has a `destroy` method which unfortunately is async. How would we address that? Im thinking on disposing SessionController both RtcEngine and SessionController just because we can't make SessionController's dispose async.

**Comment by tadaspetra** [comment]
> @tadaspetra I've found out that RtcEngine has a `destroy` method which unfortunately is async. How would we address that? Im thinking on disposing both RtcEngine and SessionController on AgoraClient's `dispose` just because we can't make SessionController's dispose async.

So that might work, but one thing to look out for in the `endCall()` function we are already calling `engine.destroy`. I'm thinking this should be done in a dispose as well somewhere in case someone leaves without "ending the call" (maybe using back button or something). 

It might be a little more complex solution for this.

File: example/lib/main.dart (hunk 2 above)
Comment (maxxfrazer) [reply example/lib/main.dart]: Very true, it's just the example project, not the actual library.

Comment (Ascenio) [reply example/lib/main.dart]: Then it's done :smiley:

**Comment by tadaspetra** [comment]
Thank you for the idea, and your code was really useful in the merge of #45. 

Thank you for all the contributions and hope to see you around here more :)

**Pull request closed without merging**